### PR TITLE
Cache Aware Streaming tutorial notebook

### PR DESCRIPTION
**Pull request by erastorgueva-nv**
# What does this PR do ?

Adding the notebook that allows us to do real-time streaming using a cache-aware fastconformer model (supports single-lookahead and multi-lookahead models).

**Collection**: `tutorials/asr`

# Changelog 
- Add cache-aware streaming notebook in `tutorials/asr`
- Add call to `self.setup_streaming_params` inside `set_default_att_context_size` method of `ConformerEncoder`.

# Usage

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Changes were originally made in https://github.com/NVIDIA/NeMo/pull/8036 but that PR was closed, replaced with this one.

### CODE REVIEW

**Comment by erastorgueva-nv** [comment]
jenkins

**Comment by titu1994** [comment]
Notebooks don't need to pass Jenkins ~ if it's reviewed and approved I can merge

File: docs/source/asr/intro.rst [hunk 1]
```diff
@@ -109,7 +109,7 @@ Use real-time transcription
 ---------------------------
 
 It is possible to use NeMo to transcribe speech in real-time. You can find an example of how to do 
-this in the following `notebook tutorial <https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Online_ASR_Microphone_Demo.ipynb>`_.
+this in the following `notebook tutorial <https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb>`_.
```
Comment (VahidooX) [inline docs/source/asr/intro.rst:112]: Let's add both tutorials here.

File: docs/source/starthere/tutorials.rst [hunk 2]
```diff
@@ -47,8 +47,11 @@ To run a tutorial:
      - Offline ASR Inference with Beam Search and External Language Model Rescoring
      - `Offline ASR <https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/Offline_ASR.ipynb>`_
    * - ASR
-     - Online ASR inference with Microphone
-     - `Online ASR Microphone <https://github.com/NVIDIA/NeMo/blob/stable/tutorials/asr/Online_ASR_Microphone_Demo.ipynb>`_
+     - Online ASR inference with Microphone (Cache-Aware Streaming)
+     - `Online ASR Microphone <https://github.com/NVIDIA/NeMo/blob/stable/tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb>`_
```
Comment (VahidooX) [inline docs/source/starthere/tutorials.rst:51]: Titles of both links are the same "Online ASR Microphone".

**Review by VahidooX — DISMISSED** [review]
Just left some minor comments.

**Review by VahidooX — DISMISSED** [review]

File: tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    "if model_name == \"stt_en_fastconformer_hybrid_large_streaming_multi\":\n",
+    "    # check that lookahead_size is one of the valid ones\n",
+    "    if lookahead_size not in [0, 80, 480, 1040]:\n",
+    "        raise ValueError(\n",
+    "            f\"specified lookahead_size {lookahead_size} is not one of the \"\n",
+    "            \"allowed lookaheads (can select 0, 80, 480 or 1040 ms)\"\n",
+    "        )\n",
+    "\n",
+    "    # update att_context_size\n",
+    "    asr_model.encoder.set_default_att_context_size([70, int(lookahead_size / ENCODER_STEP_LENGTH)])\n",
```
Comment (titu1994) [inline tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb:184]: Don't hard code 70, try to pull from config, maybe from encoder config ?

File: tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+   "source": [
+    "# init params we will use for streaming\n",
+    "previous_hypotheses = None\n",
+    "pred_out_stream = None\n",
+    "step_num = 0\n",
+    "pre_encode_cache_size = asr_model.encoder.streaming_cfg.pre_encode_cache_size[1]\n",
+    "# cache-aware models require some small section of the previous processed_signal to\n",
+    "# be fed in at each timestep - we initialize this to a tensor filled with zeros\n",
+    "# so that we will do zero-padding for the very first chunk(s)\n",
+    "cache_pre_encode = torch.zeros((1, 80, pre_encode_cache_size), device=asr_model.device)\n",
```
Comment (titu1994) [inline tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb:237]: Read the 80 from config preprocessor

**Review by titu1994 — COMMENTED** [review]
Overall looks very good, minor comments

File: docs/source/asr/models.rst [hunk 5]
```diff
@@ -159,6 +159,8 @@ You may find more examples under ``<NeMo_git_root>/examples/asr/conf/fastconform
 Cache-aware Streaming Conformer
 -------------------------------
 
+Try real-time ASR with the Cache-aware Streaming Conformer `tutorial notebook <https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Online_ASR_Microphone_Demo_Cache_Aware_Streaming.ipynb>`_.
```
[the remark this answers is not in the document]
Comment (erastorgueva-nv) [reply docs/source/asr/models.rst:162]: Not sure what you mean? PR is to merge into 1.23, so we are trying to make sure the notebook will be available in 1.23

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 6]
```diff
@@ -786,6 +786,8 @@ def set_default_att_context_size(self, att_context_size):
         if att_context_size is not None:
             self.att_context_size = att_context_size
 
+        self.setup_streaming_params()
```
[the remark this answers is not in the document]
Comment (erastorgueva-nv) [reply nemo/collections/asr/modules/conformer_encoder.py:789]: Yes, I believe it is safe to call it, Vahid recommended me to add it there to fix a bug we were facing with the multi-lookahead models (cc @VahidooX )

File: docs/source/asr/models.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply docs/source/asr/models.rst:162]: Ah ok. I didn't notice it was to 1.23. Disregard comment

**Review by VahidooX — APPROVED** [review]

**Review by titu1994 — APPROVED** [review]
Looks great !

**Pull request merged**